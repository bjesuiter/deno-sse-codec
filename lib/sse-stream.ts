import { SSEEvent } from "./types/sse-event.ts";
import { IllegalArgumentError } from "./types/illegal-argument-error.ts";
import { encodeSSEEvent } from "./encode-sse-event.ts";

/**
 * A class to encapsulate an SSEStream which can be used to read from it as readable stream or write to it.
 */
export class SSEStream extends ReadableStream<SSEEvent> {
  private controller?: ReadableStreamDefaultController;

  constructor() {
    super({
      start: (controller) => {
        this.controller = controller;
      },
      pull: () => {},
      cancel: () => {},
    });
  }

  /** Enqeues sseEvents on demand! */
  public emit(event: SSEEvent) {
    if (!this.controller) {
      throw new IllegalArgumentError(
        `No readable stream controller available yet. Please wait a bit longer, it should appear automatically`,
      );
    }

    const sseString = encodeSSEEvent(event);
    this.controller.enqueue(sseString);
  }

  public emitMany(events: SSEEvent[]) {
    if (!this.controller) {
      throw new IllegalArgumentError(
        `No readable stream controller available yet. Please wait a bit longer, it should appear automatically`,
      );
    }

    for (const event of events) {
      const sseString = encodeSSEEvent(event);
      this.controller.enqueue(sseString);
    }
  }

  /**
   * Closes the underlying controller of this SSEStream
   */
  public end() {
    this.controller?.close();
  }
}
