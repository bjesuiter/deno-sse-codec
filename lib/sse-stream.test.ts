import { SSEStream } from "./sse-stream.ts";
import {
  assertExists,
  assertSnapshot,
  describe,
  it,
} from "../dependencies/_testing.std.ts";

describe(`sse-stream`, () => {
  it(`can be constructed`, () => {
    const sseStream = new SSEStream();
    assertExists(sseStream);
  });

  it(`can be ended`, () => {
    const sseStream = new SSEStream();
    sseStream.end();
  });

  it(`can receive events from readableStream property`, async (t) => {
    const sseStream = new SSEStream();
    const reader = sseStream.readableStream.getReader();

    sseStream.emit({ name: "event1" });
    sseStream.emit({ name: "event2" });

    const chunk1 = await reader.read();
    const chunk2 = await reader.read();

    console.log(chunk1);
    console.log(chunk2);

    assertSnapshot(t, [chunk1, chunk2]);
    sseStream.end();
  });
});
