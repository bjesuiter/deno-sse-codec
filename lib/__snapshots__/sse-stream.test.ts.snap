export const snapshot = {};

snapshot[`sse-stream > can receive events from readableStream property 1`] = `
[
  {
    done: false,
    value: "event: event1 

",
  },
  {
    done: false,
    value: "event: event2 

",
  },
]
`;
